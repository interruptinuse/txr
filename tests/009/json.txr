@;
@; A JSON value is a string, number, associative object, keyword or array.
@;
@(define value (v))@\
  @(cases)@\
    @(string v)@(or)@(num v)@(or)@(object v)@(or)@\
    @(keyword v)@(or)@(array v)@\
  @(end)@\
@(end)
@;
@; Pattern function for matching whitespace
@;
@(define ws)@/[\n\t ]*/@(end)
@;
@; Pattern function for matching a JSON string, with all the
@; escape sequences.
@;
@(define string (s))@\
  @(local hex)@\
  @(ws)@\
  "@(coll :gap 0 :vars (s))@\
     @(cases)@\
       \"@(bind s "&quot;")@(or)@\
       \\@(bind s "\\\\")@(or)@\
       \/@(bind s "\\/")@(or)@\
       \b@(bind s "&#8;")@(or)@\
       \f@(bind s "&#12;")@(or)@\
       \n@(bind s "&#10;")@(or)@\
       \r@(bind s "&#13;")@(or)@\
       \t@(bind s "&#9;")@(or)@\
       \u@{hex /[0-9A-Fa-f][0-9A-Fa-f][0-9A-Fa-f][0-9A-Fa-f]/}@\
         @(bind s `&#x@hex;`)@(or)@\
       @{s /[^"\\]*/}@(filter :to_html s)@\
     @(end)@\
     @(until)"@\
   @(end)"@\
  @(ws)@\
  @(cat s "")@\
  @(filter :from_html s)@\
@(end)
@;
@; Pattern function for recognizing a number.
@;
@(define num (n))@\
  @(local tok)@\
  @(ws)@{tok /-?[0-9]+((\.[0-9]+)?([Ee][+\-]?[0-9]+)?)?/}@(ws)@\
  @(bind n @(flo-str tok))@\
@(end)
@;
@; Recognize the JSON keyword true, false and null, turning
@; them into TXR Lisp keywords
@;
@(define keyword (k))@\
  @(local tok)@\
  @(all)@(ws)@{tok /true|false|null/}@(trailer)@/[^A-Za-z0-9_]/@(end)@(ws)@\
  @(bind k @(intern tok keyword-package))@\
@(end)
@;
@; Recognize an object: a collection of string/value pairs,
@; turning them into a list.
@;
@(define object (v))@\
  @(local p e pair)@\
  @(ws){@(ws)@(coll :gap 0 :vars (v))@\
                @(string p):@(value e)@/,?/@\
                @(bind v (p e))@\
                @(until)}@\
             @(end)}@(ws)@\
@(end)
@;
@; Recognize an array.
@;
@(define array (v))@\
  @(local e)@\
  @(ws)[@(ws)@(coll :gap 0 :vars (e))@(value e)@/,?/@(until)]@(end)]@(ws)@\
  @(bind v @(progn ^#(,*e)))@\
@(end)
@;
@; Now parse the input as a JSON object
@;
@(next :args)
@(collect)
@file
@(next file)
@(freeform)
@(maybe)@(value ast)@(end)@badsyntax
@;
@; Output resulting abstract syntax tree.
@;
@(do (format t "AST: ~s\n\n" ast)
     (format t "Unmatched junk: ~s\n\n" badsyntax))
@(end)
