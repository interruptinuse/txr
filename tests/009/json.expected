AST: (("web-app" (("servlet" #((("servlet-name" "cofaxCDS") ("servlet-class" "org.cofax.cds.CDSServlet")
                                ("init-param" (("configGlossary:installationAt" "Philadelphia, PA") ("configGlossary:adminEmail" "ksm@pobox.com")
                                               ("configGlossary:poweredBy" "Cofax") ("configGlossary:poweredByIcon" "/images/cofax.gif")
                                               ("configGlossary:staticPath" "/content/static") ("templateProcessorClass" "org.cofax.WysiwygTemplate")
                                               ("templateLoaderClass" "org.cofax.FilesTemplateLoader") ("templatePath" "templates")
                                               ("templateOverridePath" "") ("defaultListTemplate" "listTemplate.htm")
                                               ("defaultFileTemplate" "articleTemplate.htm") ("useJSP" :false)
                                               ("jspListTemplate" "listTemplate.jsp") ("jspFileTemplate" "articleTemplate.jsp")
                                               ("cachePackageTagsTrack" 200.0) ("cachePackageTagsStore" 200.0)
                                               ("cachePackageTagsRefresh" 60.0) ("cacheTemplatesTrack" 100.0)
                                               ("cacheTemplatesStore" 50.0) ("cacheTemplatesRefresh" 15.0) ("cachePagesTrack" 200.0)
                                               ("cachePagesStore" 100.0) ("cachePagesRefresh" 10.0) ("cachePagesDirtyRead" 10.0)
                                               ("searchEngineListTemplate" "forSearchEnginesList.htm") ("searchEngineFileTemplate" "forSearchEngines.htm")
                                               ("searchEngineRobotsDb" "WEB-INF/robots.db") ("useDataStore" :true)
                                               ("dataStoreClass" "org.cofax.SqlDataStore") ("redirectionClass" "org.cofax.SqlRedirection")
                                               ("dataStoreName" "cofax") ("dataStoreDriver" "com.microsoft.jdbc.sqlserver.SQLServerDriver")
                                               ("dataStoreUrl" "jdbc:microsoft:sqlserver://LOCALHOST:1433;DatabaseName=goon")
                                               ("dataStoreUser" "sa") ("dataStorePassword" "dataStoreTestQuery")
                                               ("dataStoreTestQuery" "SET NOCOUNT ON;select test='test';") ("dataStoreLogFile" "/usr/local/tomcat/logs/datastore.log")
                                               ("dataStoreInitConns" 10.0) ("dataStoreMaxConns" 100.0) ("dataStoreConnUsageLimit" 100.0)
                                               ("dataStoreLogLevel" "debug") ("maxUrlLength" 500.0))))
                               (("servlet-name" "cofaxEmail") ("servlet-class" "org.cofax.cds.EmailServlet")
                                ("init-param" (("mailHost" "mail1") ("mailHostOverride" "mail2"))))
                               (("servlet-name" "cofaxAdmin") ("servlet-class" "org.cofax.cds.AdminServlet"))
                               (("servlet-name" "fileServlet") ("servlet-class" "org.cofax.cds.FileServlet"))
                               (("servlet-name" "cofaxTools") ("servlet-class" "org.cofax.cms.CofaxToolsServlet")
                                ("init-param" (("templatePath" "toolstemplates/") ("log" 1.0) ("logLocation" "/usr/local/tomcat/logs/CofaxTools.log")
                                               ("logMaxSize" "") ("dataLog" 1.0) ("dataLogLocation" "/usr/local/tomcat/logs/dataLog.log")
                                               ("dataLogMaxSize" "") ("removePageCache" "/content/admin/remove?cache=pages&id=")
                                               ("removeTemplateCache" "/content/admin/remove?cache=templates&id=")
                                               ("fileTransferFolder" "/usr/local/tomcat/webapps/content/fileTransferFolder")
                                               ("lookInContext" 1.0) ("adminGroupID" 4.0) ("betaServer" :true))))))
                  ("servlet-mapping" (("cofaxCDS" "/") ("cofaxEmail" "/cofaxutil/aemail/*") ("cofaxAdmin" "/admin/*")
                                      ("fileServlet" "/static/*") ("cofaxTools" "/tools/*")))
                  ("taglib" (("taglib-uri" "cofax.tld") ("taglib-location" "/WEB-INF/tlds/cofax.tld"))))))

Unmatched junk: ""

AST: #("JSON Test Pattern pass1" (("object with 1 member" #("array with 1 element")))
       nil #() -42.0 :true :false :null (("integer" 1234567890.0) ("real" -9876.54321) ("e" 1.23456789e-13)
                                         ("E" 1.23456789e34) ("" 2.3456789012e76) ("zero" 0.0) ("one" 1.0)
                                         ("space" " ") ("quote" "\"") ("backslash" "\\\\") ("controls" "\b\f\n\r\t")
                                         ("slash" "/ & \\/") ("alpha" "abcdefghijklmnopqrstuvwyz") ("ALPHA" "ABCDEFGHIJKLMNOPQRSTUVWYZ")
                                         ("digit" "0123456789") ("0123456789" "digit") ("special" "`1~!@#$%^&*()_+-={':[,]}|;.</>?")
                                         ("hex" "ģ䕧覫췯ꯍ") ("true" :true) ("false" :false) ("null" :null)
                                         ("array" #()) ("object" nil) ("address" "50 St. James Street")
                                         ("url" "http://www.JSON.org/") ("comment" "// /* <!-- --") ("# -- --> */" " ")
                                         (" s p a c e d " #(1.0 2.0 3.0 4.0 5.0 6.0 7.0)) ("compact" #(1.0 2.0 3.0 4.0 5.0 6.0 7.0))
                                         ("jsontext" "{\"object with 1 member\":[\"array with 1 element\"]}")
                                         ("quotes" "&#34; \" %22 0x22 034 &#x22;") ("\\/\\\\\"쫾몾ꮘﳞ볚\b\f\n\r\t`1~!@#$%^&*()_+-=[]{}|;:',./<>?"
                                                                                    "A key can be any string"))
       0.5 98.6 99.44 1066.0 10.0 1.0 0.1 1.0 2.0 2.0 "rosebud")

Unmatched junk: ""

